tgui/packages/tgui/interfaces/SolarControl.tsx

```diff
@@ -1,66 +1,117 @@
+import { BooleanLike } from 'common/react';
 import { useBackend } from '../backend';
-import { Box, Button, Grid, LabeledList, NumberInput, ProgressBar, Section } from '../components';
+import { Box, Button, LabeledList, NumberInput, ProgressBar, Chart, Section, Stack } from '../components';
 import { Window } from '../layouts';
 
+type Data = {
+  supply: number;
+  capacity: number;
+  azimuth_current: number;
+  azimuth_rate: number;
+  max_rotation_rate: number;
+  tracking_state: number;
+  connected_panels: number;
+  connected_tracker: BooleanLike;
+  history: History;
+};
+
+type History = {
+  supply: number[];
+  capacity: number[];
+};
+
 export const SolarControl = (props, context) => {
-  const { act, data } = useBackend(context);
+  const { act, data } = useBackend<Data>(context);
   const {
-    generated,
-    generated_ratio,
+    supply,
+    capacity,
     azimuth_current,
     azimuth_rate,
     max_rotation_rate,
     tracking_state,
     connected_panels,
     connected_tracker,
+    history,
   } = data;
+  const supplyData = history.supply.map((value, i) => [i, value]);
+  const capacityData = history.capacity.map((value, i) => [i, value]);
```

Comment:
```suggestion
  const supplyData = history?.supply?.map((value, i) => [i, value]);
  const capacityData = history?.capacity?.map((value, i) => [i, value]);
```
no guarantee these are sent. other option is to destructure and cast,
```
const { capacity = [], supply = [] } = history;
```

Reply:
They are guaranteed to be sent. I'd say it should throw an exception if they're not.
![image](https://github.com/tgstation/tgstation/assets/3625094/d0b4e538-2053-418c-96b0-af82ab4a182e)

Reply:
it'll give the ntos error, but why block the ui on this? "no guarantee these are sent" is not an indictment of your code btw

Reply:
If it gives error, then it means that someone broke it. Right now it sends the empty arrays when there is no history and UI handles it properly. The screenshot above is the payload from backend when there is no data.

Reply:
the backend is possibly undefined :)
calls to and from another service will always be a yeet into/from the void

Reply:
Then the entire UI will be broken, right? Not just history. I'm not aware of such tricky logic of SS13 UI, but I try to avoid `?`s as much as possible, because if the null pointer happened, then it's a sign that it requires fixing. Sorry for being stubborn here.

Reply:
no apologies needed, i understand your reasoning. different schools of thought, you want it to display an error if something went wrong, I want it to persist as if it didn't happen. the case is highly unlikely, so it doesn't warrant stopping the PR